### PR TITLE
 Store the current config file at read time, and return it from GetConfigFile if set

### PR DESCRIPTION
**Pull request by Empact**
This maintains accuracy in the context of `datadirectory` changes,
e.g. #12722

Also drop the related arguments to `GetConfigFile` and `ReadConfigFile`, and follow the example of `GetDebugLogPath` by calling `GetArg` within the `GetConfigFile` function.

### CODE REVIEW

**Review by ryanofsky — COMMENTED** [review]
utACK f73aed2d78391bcf8aeaa16e69f72d74e9dfb163. This is a nice fix for #12722 and also seems like a good code simplification.

**Comment by Empact** [comment]
Added some testing for `ReadConfigFile`, which wasn't previously covered.

**Comment by MarcoFalke** [comment]
Haven't looked closely, but this seems to revert #8856

File: src/test/getarg_tests.cpp [hunk 1]
```diff
@@ -30,6 +30,41 @@ static void ResetArgs(const std::string& strArg)
...
 
+BOOST_AUTO_TEST_CASE(configfile)
+{
+  std::ostringstream args;
+  args << "-conf=" << (fs::path("data") / "bitcoin.conf").string()
+       << " -datadir=" << fs::path("src/test").string();
+  ResetArgs(args.str());
+  gArgs.ReadConfigFile();
+
+  BOOST_CHECK_EQUAL(gArgs.GetConfigFile(), fs::system_complete("test/data/bitcoin.conf"));
```
Comment (MarcoFalke) [inline src/test/getarg_tests.cpp]:
```
test/getarg_tests.cpp(41): error: in 
"getarg_tests/configfile": check gArgs.GetConfigFile() == fs::system_complete("test/data/bitcoin.conf") has failed
["/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/src/data/bitcoin.conf" != "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/src/test/data/bitcoin.conf"]

Comment (Empact) [reply src/test/getarg_tests.cpp]: Yeah it's running different locally for me than on CI, and have not figured that out yet. In cases where the paths match up as expected (which should mean the config is properly identified), the arg checks do not pass, as though the file is not being picked up.

Comment (Empact) [reply src/test/getarg_tests.cpp]: Any idea what's going on here? https://travis-ci.org/bitcoin/bitcoin/jobs/358040567

It's saying the test bitcoin.conf does not exist under CI:

```
test/getarg_tests.cpp(39): error in "configfile": check fs::exists(gArgs.GetConfigFile()) failed
test/getarg_tests.cpp(40): error in "configfile": check gArgs.GetConfigFile() == "test/data/bitcoin.conf" failed ["/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/src/test/data/bitcoin.conf" != test/data/bitcoin.conf]
```

**Comment by Empact** [comment]
Alternate title for this PR: "In which do not RTFM and instead discover the existence of Makefile.test.include the hard way"

**Comment by Empact** [comment]
Re #8856, seems the file has gone the other direction since then - maybe due to the introduction of ArgsManager.

Currently the following other util methods use GetArg:
* GetDataDir
* GetPidFile
* GetDebugLogPath

In this case seems the change is required if we want to retain the state.

**Review by ryanofsky — COMMENTED** [review]
utACK 4692fc01003e39719c50632ba16339fd5173a856. Only difference since last review is adding tests (thanks!), declaring `GetConfigFile` const, and some formatting changes.

#8856 doesn't match current organization of the code not so relevant. Following that approach seems more repetitive, and also like it would prevent fixing the bug here.

**Review by ryanofsky — COMMENTED** [review]
utACK 900e3e54890ed3d2c9ba3362794c928bfdc90151. Just rebase, formatting changes, and new config file test since last review.

This PR depends on #12812 which still has some comments that need to be addressed there.

**Comment by Empact** [comment]
Thinking to hold this pending #12878

**Comment by MarcoFalke** [comment]
Needs rebase if still relevant

**Comment by Empact** [comment]
Rebased and updated tests to package the config file via the `TEST_FILES_*` setup.

**Comment by ajtowns** [comment]
Two thoughts:
 * rather than have GetConfigFile use gArgs, probably better to make GetConfigFile a member of ArgsManager?
 * might make sense to use to split between config args and command line args in #11862 and have GetConfigFile/ReadConfigFile only lookup command line args for the value for "-conf", rather than saving it separately?

**Comment by MarcoFalke** [comment]
Sorry, needs rebase again.

**Pull request closed without merging**